scripts/components/api-changes-validator/usage_statemets_renderer.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  NamespaceDefinitions,
+  UsageStatementsGeneratorOutput,
+} from './types.js';
+import { EOL } from 'os';
+
+/**
+ * Renders usage statements.
+ */
+export class UsageStatementsRenderer {
+  /**
+   * Creates import statements renderer.
+   */
+  constructor(
+    private readonly generatorOutputs: Array<UsageStatementsGeneratorOutput>,
+    private readonly namespaceDefinitions: NamespaceDefinitions
+  ) {}
+  render = (): string => {
+    let usageStatements = this.renderCombinedUsageStatements();
+    usageStatements = this.applyNamespaces(usageStatements);
+    return usageStatements;
+  };
+
+  private renderCombinedUsageStatements = (): string => {
+    const usageStatements: Array<string> = [];
+    for (const generatorOutput of this.generatorOutputs) {
+      if (generatorOutput.usageStatement) {
+        usageStatements.push(generatorOutput.usageStatement);
+      }
+    }
+    return usageStatements.join(EOL);
+  };
+
+  /**
+   * This function applies namespaces using a simple algorithm that
+   * regex match and replace symbols in generated code with namespace prefix.
+   *
+   * This a cheap alternative to a proper but more expensive solution
+   * that would involve inspecting symbols during generation.
+   * Proper solution is expensive because we'd need to unpack every statement
+   * that can contain these types like generic type arguments, function arguments, etc.
+   * We do leverage fact that these snippets can be copied over from API extract verbatim in generators to simplify them.
+   * Proper solution should be considered when this approach no longer works.
+   */
+  private applyNamespaces = (usageStatements: string): string => {
+    for (const symbolName of this.namespaceDefinitions.namespaceBySymbol.keys()) {
+      if (this.namespaceDefinitions.namespaceNames.has(symbolName)) {
+        // skip namespaces
+        continue;
+      }
+      let currentSymbolName: string | undefined = symbolName;
+      const namespaceHierarchy: Array<string> = [];
+      do {
+        currentSymbolName =
+          this.namespaceDefinitions.namespaceBySymbol.get(currentSymbolName);
+        if (currentSymbolName) {
+          namespaceHierarchy.unshift(currentSymbolName);
+        }
+      } while (currentSymbolName);
+      const symbolWithNamespace = `${namespaceHierarchy.join(
+        '.'
+      )}.${symbolName}`;
+
+      // characters that can be found before or after symbol
+      // this is to prevent partial matches in case one symbol's characters are subset of longer one
+      const symbolTerminators = '[\\s\\,\\(\\)<>]';
```

Comment:
Do we try to match any charactors in `\s,()<>`?

Would it be helpful to make a comment about what exactly we try to match here with the Regex?

Reply:
> Do we try to match any charactors in \s,()<>?

Correct.

> Would it be helpful to make a comment about what exactly we try to match here with the Regex?

"characters that can be found before or after symbol" in the comment above is indicating that. In TS/JS there's finite set of characters that can be found around symbols, i.e. whitespace, comma and bracket. 
This set might not be complete right now, but can be expanded when new patterns are added to API.md files.